src/main/java/ua/chernonog/onlinebookstore/service/impl/CartItemServiceImpl.java

```diff
@@ -0,0 +1,60 @@
+package ua.chernonog.onlinebookstore.service.impl;
+
+import jakarta.transaction.Transactional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import ua.chernonog.onlinebookstore.dto.request.cartitem.CartItemRequestDto;
+import ua.chernonog.onlinebookstore.entity.CartItem;
+import ua.chernonog.onlinebookstore.entity.ShoppingCart;
+import ua.chernonog.onlinebookstore.exception.DuplicateEntityException;
+import ua.chernonog.onlinebookstore.exception.EntityNotFoundException;
+import ua.chernonog.onlinebookstore.mapper.CartItemMapper;
+import ua.chernonog.onlinebookstore.repository.book.BookRepository;
+import ua.chernonog.onlinebookstore.repository.cartitem.CartItemRepository;
+import ua.chernonog.onlinebookstore.service.CartItemService;
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Transactional
+    @Override
+    public CartItem save(
+            CartItemRequestDto cartItemRequestDto,
+            ShoppingCart shoppingCartFromDb
+    ) {
+        cartItemRepository.findBookInCart(
+                        cartItemRequestDto.getBookId(),
+                        shoppingCartFromDb.getId()
+                )
+                .ifPresent(s -> {
+                    throw new DuplicateEntityException("This book has already been added");
+                });
+
+        CartItem cartItem = cartItemMapper.toModel(cartItemRequestDto);
+        cartItem.setBook(bookRepository.findById(cartItemRequestDto.getBookId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find book with id " + cartItemRequestDto.getBookId())));
+        cartItem.setShoppingCart(shoppingCartFromDb);
```

Comment:
Code duplication, you should have bookRepository.findById in bookService, so I suggest to inject book service here and use it

Reply:
This question was on QnA with Eugene, if you add service, you will need to add Book Mapper. It was decided to simply refer directly to the repository